File: sport-log-server/src/types/movement.rs
use diesel_derive_enum::DbEnum;
use serde::{Deserialize, Serialize};

use sport_log_server_derive::{
    Create, Delete, GetAll, GetById, InnerIntFromSql, InnerIntToSql, Update,
};

use crate::{
    schema::{eorm, movement},
    types::UserId,
};

#[derive(DbEnum, Debug, Serialize, Deserialize)]
pub enum MovementCategory {
    Cardio,
    Strength,
}

#[derive(DbEnum, Debug, Serialize, Deserialize)]
pub enum MovementUnit {
    Reps,
    Cal,
    Meter,
    Km,
    Yard,
    Foot,
    Mile,
}

#[derive(
    FromSqlRow,
    AsExpression,
    Serialize,
    Deserialize,
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    InnerIntToSql,
    InnerIntFromSql,
)]
#[sql_type = "diesel::sql_types::Integer"]
pub struct MovementId(pub i32);

#[derive(
    Queryable, AsChangeset, Serialize, Deserialize, Debug, Create, GetById, GetAll, Update, Delete,
)]
#[table_name = "movement"]
pub struct Movement {
    pub id: MovementId,
    pub user_id: UserId,
    pub name: String,
    pub description: Option<String>,
    pub category: MovementCategory,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name = "movement"]
pub struct NewMovement {
    pub user_id: UserId,
    pub name: String,
    pub description: Option<String>,
    pub category: MovementCategory,
}

#[derive(
    FromSqlRow,
    AsExpression,
    Serialize,
    Deserialize,
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    InnerIntToSql,
    InnerIntFromSql,
)]
#[sql_type = "diesel::sql_types::Integer"]
pub struct EormId(pub i32);

#[derive(
    Queryable, AsChangeset, Serialize, Deserialize, Debug, Create, GetById, GetAll, Update, Delete,
)]
#[table_name = "eorm"]
pub struct Eorm {
    pub id: EormId,
    pub reps: i32,
    pub percentage: f32,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name = "eorm"]
pub struct NewEorm {
    pub reps: i32,
    pub percentage: f32,
}

File: sport-log-client/frontend/lib/models/movement.dart

enum MovementUnit {
  reps, cal, meter, km, yard, foot, mile
}

extension ToDisplayName on MovementUnit {
  String toDisplayName() {
    switch (this) {
      case MovementUnit.reps:
        return "Reps";
      case MovementUnit.cal:
        return "Cals";
      case MovementUnit.meter:
        return "m";
      case MovementUnit.km:
        return "km";
      case MovementUnit.yard:
        return "yd";
      case MovementUnit.foot:
        return "ft";
      case MovementUnit.mile:
        return "mi";
    }
  }
}

enum MovementCategory {
  strength, cardio
}

class NewMovement {
  NewMovement({
    required this.name,
    required this.category,
    this.description,
  });

  String name;
  MovementCategory category;
  String? description;
}

class Movement {
  Movement({
    required this.id,
    required this.userId,
    required this.name,
    required this.category,
    this.description,
  });

  int id;
  int? userId;
  String name;
  MovementCategory category;
  String? description;

  Movement.fromNewMovement(NewMovement nm, this.id, this.userId)
    : name = nm.name, description = nm.description, category = nm.category;
}